core/src/avm1/globals/string.rs

```diff
@@ -439,7 +439,11 @@ fn substr<'gc>(
 
     let len = match args.get(1) {
         None | Some(Value::Undefined) => this_len,
-        Some(n) => string_index(n.coerce_to_i32(activation)?, this_len),
+        Some(n) => string_index_substr(
+            args.get(0).unwrap().coerce_to_i32(activation)?,
```

Comment:
Because `coerce_to_i32` can have a side effect (e.g. an object with a `valueOf` function), let's store `args.get(0).coerce_to_i32` in a variable and use it both in the `string_wrapping_index` call and the `string_index_substr` call, instead of recalculating it.